docs/style-guide.md

```diff
@@ -0,0 +1,291 @@
+    TranslateScaleRotate
+}
+
+impl Default for TransformOrder {
+    fn default() -> Self { Self::ScaleRotateTranslate }
+}
+
+
+// =============================
+// === HierarchicalTransform ===
+// =============================
+
+pub struct HierarchicalTransform<OnChange> {
+    transform        : Transform,
+    transform_matrix : Matrix4<f32>,
+    origin           : Matrix4<f32>,
+    matrix           : Matrix4<f32>,
+    pub dirty        : dirty::SharedBool<OnChange>,
+    pub logger       : Logger,
+}
+
+impl<OnChange> HierarchicalTransform<OnChange> {
+    pub fn new(logger:Logger, on_change:OnChange) -> Self {
+        let logger_dirty     = logger.sub("dirty");
+        let transform        = default();
+        let transform_matrix = Matrix4::identity();
+        let origin           = Matrix4::identity();
+        let matrix           = Matrix4::identity();
+        let dirty            = dirty::SharedBool::new(logger_dirty,on_change);
+        Self {transform,transform_matrix,origin,matrix,dirty,logger}
+    }
+}
+
+// === Getters ===
+
+impl<OnChange> HierarchicalTransform<OnChange> {
+    pub fn position(&self) -> &Vector3<f32> {
+        &self.transform.position
+    }
+
+    pub fn rotation(&self) -> &Vector3<f32> {
+        &self.transform.rotation
+    }
+
+    ...
+}
+
+// === Setters ===
+
+impl<OnChange:Callback0> HierarchicalTransform<OnChange> {
+    pub fn position_mut(&mut self) -> &mut Vector3<f32> {
+        self.dirty.set();
+        &mut self.transform.position
+    }
+
+    pub fn rotation_mut(&mut self) -> &mut Vector3<f32> {
+        self.dirty.set();
+        &mut self.transform.rotation
+    }
+
+    ...
+}
+```
+
+
+### Vertical spacing
+We use the following amount of vertical spacing:
+- 2 blank lines after imports
+- 2 blank lines before each section.
+- 1 blank line before and after sub-section.
+- 1 blank line before functions / structures / impls.
+- 1 blank line at the end of the file.
+
+
+### Multiline Expressions
+Most (preferably all) expressions should be single line. Multiline expression is
+hard to read and introduces noise in the code. Often, it is also an indicator of
+a code which is not properly refactored. Try to refactor parts of multiline 
+expressions to well-named variables, and divide them to several single-line 
+expressions.
+
+Example of bad formatted code:
+```rust
+pub fn new() -> Self {
+    let shape_dirty = ShapeDirty::new(logger.sub("shape_dirty"),
+        on_dirty.clone());
+    let dirty_flag = MeshRegistryDirty::new(logger.sub("mesh_registry_dirty"),
+        on_dirty);
+    Self { dirty_flag, dirty_flag }
+}
+```
+Example of properly formatted code:
+
+```rust
+pub fn new() -> Self {
+    let sub_logger  = logger.sub("shape_dirty");
+    let shape_dirty = ShapeDirty::new(sub_logger,on_dirty.clone());
+    let sub_logger  = logger.sub("mesh_registry_dirty");
+    let dirty_flag  = MeshRegistryDirty::new(sub_logger,on_dirty);
+    Self {shape_dirty,dirty_flag}
+}
+```
```

Comment:
I'd like to put here a case we discussed on call (before Michał and Danilo joined): a chain of processing functions, characteristic for iterators. Compare:
```rust
impl<'a> CharVerticesBuilder<'a> {
    // ........
    pub fn build(&mut self, character : char) -> OneCharVertices {
        // a bunch if lets computing ultimate transformation
        let pair_iter            = GLYPH_VERTICES_BASE_LAYOUT.iter();
        let point_iter           = pair_iter.map(|(x, y)| Point2::new(*x, *y));
        let transformed          = point_iter.map(|p| ultimate_vertex_trans * p);
        let transformed_iterable = transformed.map(|p| point_to_iterable(p));
        transformed_iterable.flatten().collect()
    }
```
with
```rust
impl<'a> CharVerticesBuilder<'a> {
    // ........
    pub fn build(&mut self, character : char) -> OneCharVertices {
        // a bunch if lets computing ultimate transformation
        GLYPH_VERTICES_BASE_LAYOUT.iter()
            .map(|(x, y)| Point2::new(*x, *y))
            .map(|p| ultimate_vertex_trans * p)
            .map(|p| point_to_iterable(p))
            .flatten().collect()
    }
```
For me the second case is much less noisy. It's hard to make a meaningful label here.

And, unfortunately I still think same about optional example we discussed:
```rust
    fn get_kerning_trans(&mut self, ch : char) -> Translation2<f32> {
        let prev_char = self.previous_char;
        let opt_value = prev_char.map(|lc| self.font.get_kerning(lc, ch));
        let value     = opt_value.unwrap_or(0.0);
        Translation2::new(value, 0.0)
    }
```
What is the value of introduced label `opt_value`? I usually see that this is _optional_, from its type
```rust
    fn get_kerning_trans(&mut self, ch : char) -> Translation2<f32> {
        let prev_char = self.previous_char;
        let value     = prev_char.map(|lc| self.font.get_kerning(lc, ch))
                                 .unwrap_or(0.0);
        Translation2::new(value, 0.0)
    }
```
We have the same number of lines, and avoid making zero-value labels. I propose using this heavy-alignment of new function - the processing function should be a really simple call (or it actually deserve a label). 

As a rule, there should be only one processing call per line, and from the call we should easily read what the result is.

Reply:
So, after discussion with Wojtek we decided to leave _single line expressions_ rule as it is. It's hard to define accurately when the function calls chain are ok and when not, and the many let's in code are just a nuisance I can manage.